File: src_py/modules/video_splitter.py
import os
import json
import subprocess
import re
from pathlib import Path
from typing import List, Dict, Any, Optional, Tuple
from dataclasses import dataclass

from src_py.config.config import config
from src_py.utils.helpers import helpers


@dataclass
class VideoInfo:
    duration: float
    size: int
    bitrate: int
    video: Dict[str, Any]
    audio: Optional[Dict[str, Any]] = None


@dataclass
class AudioVolumeInfo:
    mean_volume: Optional[float]
    max_volume: Optional[float]
    has_audio: bool


@dataclass
class VideoSegment:
    index: int
    start_time: float
    end_time: float
    duration: float
    path: str
    filename: str


class VideoSplitter:
    def __init__(self, options: Optional[Dict[str, Any]] = None):
        options = options or {}
        self.options = {
            'segment_duration': options.get('segment_duration', config.video['default_segment_duration']),
            'output_dir': options.get('output_dir', config.temp_dir),
            **options,
        }
        self.ffmpeg_path = config.ffmpeg['path']
        self.ffprobe_path = config.ffmpeg['ffprobe_path']

    def _run_ffprobe(self, file_path: str) -> Dict[str, Any]:
        cmd = [
            self.ffprobe_path,
            '-v', 'quiet',
            '-print_format', 'json',
            '-show_format',
            '-show_streams',
            file_path
        ]
        result = subprocess.run(cmd, capture_output=True, text=True)
        if result.returncode != 0:
            raise Exception(f'ffprobe failed: {result.stderr}')
        return json.loads(result.stdout)

    def get_video_info(self, video_path: str) -> VideoInfo:
        if not os.path.exists(video_path):
            raise Exception(f'视频文件不存在: {video_path}')

        metadata = self._run_ffprobe(video_path)
        format_info = metadata.get('format', {})
        streams = metadata.get('streams', [])

        video_stream = next((s for s in streams if s.get('codec_type') == 'video'), None)
        audio_stream = next((s for s in streams if s.get('codec_type') == 'audio'), None)

        duration = float(format_info.get('duration', 0))
        size = int(format_info.get('size', 0))
        bitrate = int(format_info.get('bit_rate', 0))

        video_info = {
            'codec': video_stream.get('codec_name') if video_stream else None,
            'width': video_stream.get('width') if video_stream else None,
            'height': video_stream.get('height') if video_stream else None,
            'fps': self._parse_frame_rate(video_stream.get('avg_frame_rate')) if video_stream else config.video['default_fps'],
        }

        audio_info = None
        if audio_stream:
            audio_info = {
                'codec': audio_stream.get('codec_name'),
                'sample_rate': int(audio_stream.get('sample_rate', 0)) if audio_stream.get('sample_rate') else None,
                'channels': audio_stream.get('channels'),
            }

        return VideoInfo(
            duration=duration,
            size=size,
            bitrate=bitrate,
            video=video_info,
            audio=audio_info
        )

    def _parse_frame_rate(self, frame_rate_str: str) -> float:
        if not frame_rate_str or frame_rate_str == '0/0':
            return config.video['default_fps']
        parts = frame_rate_str.split('/')
        if len(parts) == 2:
            return float(parts[0]) / float(parts[1])
        return float(frame_rate_str)

    def split_video(self, video_path: str, options: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        options = options or {}
        video_info = self.get_video_info(video_path)
        segment_duration = options.get('segment_duration', self.options['segment_duration'])
        output_dir = options.get('output_dir', self.options['output_dir'])
        base_name = helpers.get_file_name_without_extension(video_path)

        helpers.ensure_directory(output_dir)

        segments = helpers.calculate_video_segments(video_info.duration, segment_duration)
        output_files = []

        for segment in segments:
            output_filename = f"{base_name}_segment_{str(segment['index']).zfill(3)}.mp4"
            output_path = os.path.join(output_dir, output_filename)

            self._extract_segment(video_path, output_path, segment['start_time'], segment['duration'])

            output_files.append(VideoSegment(
                index=segment['index'],
                start_time=segment['start_time'],
                end_time=segment['end_time'],
                duration=segment['duration'],
                path=output_path,
                filename=output_filename
            ))

        return {
            'original_video': video_path,
            'original_info': video_info,
            'segments': output_files,
            'segment_count': len(output_files),
        }

    def _extract_segment(self, input_path: str, output_path: str, start_time: float, duration: float):
        cmd = [
            self.ffmpeg_path,
            '-y',
            '-i', input_path,
            '-ss', str(start_time),
            '-t', str(duration),
            '-c:v', 'copy',
            '-c:a', 'copy',
            '-avoid_negative_ts', 'make_zero',
            output_path
        ]

        result = subprocess.run(cmd, capture_output=True)
        if result.returncode != 0:
            raise Exception(f'视频分割失败: {result.stderr.decode("utf-8", errors="ignore")}')

    def split_by_custom_intervals(self, video_path: str, intervals: List[Dict[str, float]], 
                                    output_dir: Optional[str] = None) -> Dict[str, Any]:
        video_info = self.get_video_info(video_path)
        output_directory = output_dir or self.options['output_dir']
        base_name = helpers.get_file_name_without_extension(video_path)

        helpers.ensure_directory(output_directory)

        output_files = []

        for i, interval in enumerate(intervals):
            start_time = interval['start_time']
            duration = interval['end_time'] - interval['start_time']

            output_filename = f"{base_name}_segment_{str(i).zfill(3)}.mp4"
            output_path = os.path.join(output_directory, output_filename)

            if start_time < 0 or start_time + duration > video_info.duration:
                raise Exception(
                    f'时间区间超出视频范围: 开始 {start_time}, 时长 {duration}, 视频总长 {video_info.duration}'
                )

            self._extract_segment(video_path, output_path, start_time, duration)

            output_files.append(VideoSegment(
                index=i,
                start_time=start_time,
                end_time=interval['end_time'],
                duration=duration,
                path=output_path,
                filename=output_filename
            ))

        return {
            'original_video': video_path,
            'original_info': video_info,
            'segments': output_files,
            'segment_count': len(output_files),
        }

    def merge_videos(self, video_paths: List[str], output_path: str, options: Optional[Dict[str, Any]] = None) -> str:
        if not video_paths:
            raise Exception('没有提供要合并的视频文件')

        options = options or {}
        concat_file = self._create_concat_file(video_paths)

        cmd = [
            self.ffmpeg_path,
            '-y',
            '-f', 'concat',
            '-safe', '0',
            '-i', concat_file,
            '-c:v', 'copy',
            '-c:a', 'copy',
            output_path
        ]

        if options.get('video_codec') and options['video_codec'] != 'copy':
            cmd[cmd.index('-c:v') + 1] = options['video_codec']
        if options.get('audio_codec') and options['audio_codec'] != 'copy':
            cmd[cmd.index('-c:a') + 1] = options['audio_codec']

        try:
            result = subprocess.run(cmd, capture_output=True)
            if result.returncode != 0:
                raise Exception(f'视频合并失败: {result.stderr.decode("utf-8", errors="ignore")}')
        finally:
            if os.path.exists(concat_file):
                os.unlink(concat_file)

        return output_path

    def _escape_path_for_ffmpeg(self, path: str) -> str:
        return path.replace("'", "'\\''")

    def _create_concat_file(self, video_paths: List[str]) -> str:
        temp_dir = config.temp_dir
        helpers.ensure_directory(temp_dir)

        concat_file_path = os.path.join(temp_dir, f"concat_{helpers.generate_unique_id()}.txt")
        content = '\n'.join([f"file '{self._escape_path_for_ffmpeg(path)}'" for path in video_paths])

        with open(concat_file_path, 'w', encoding='utf-8') as f:
            f.write(content)

        return concat_file_path

    def get_audio_volume(self, video_path: str) -> AudioVolumeInfo:
        cmd = [
            self.ffmpeg_path,
            '-i', video_path,
            '-af', 'volumedetect',
            '-f', 'null',
            '-vn',
            '-'
        ]

        result = subprocess.run(cmd, capture_output=True)
        stderr_output = result.stderr.decode('utf-8', errors='ignore')

        mean_volume_match = re.search(r'mean_volume:\s*(-?[\d.]+)', stderr_output)
        max_volume_match = re.search(r'max_volume:\s*(-?[\d.]+)', stderr_output)

        mean_volume = float(mean_volume_match.group(1)) if mean_volume_match else None
        max_volume = float(max_volume_match.group(1)) if max_volume_match else None

        return AudioVolumeInfo(
            mean_volume=mean_volume,
            max_volume=max_volume,
            has_audio=mean_volume is not None
        )

File: ts_temp/src/modules/videoSplitter.js
import ffmpeg from 'fluent-ffmpeg';
import path from 'path';
import fs from 'fs-extra';
import config from '../config/index.js';
import helpers from '../utils/helpers.js';

ffmpeg.setFfmpegPath(config.ffmpeg.path);
ffmpeg.setFfprobePath(config.ffmpeg.ffprobePath);

class VideoSplitter {
  constructor(options = {}) {
    this.options = {
      segmentDuration: options.segmentDuration || config.video.defaultSegmentDuration,
      outputDir: options.outputDir || config.directories.temp,
      ...options,
    };
  }

  async getVideoInfo(videoPath) {
    return new Promise((resolve, reject) => {
      ffmpeg.ffprobe(videoPath, (err, metadata) => {
        if (err) {
          reject(new Error(`无法获取视频信息: ${err.message}`));
          return;
        }

        const videoStream = metadata.streams.find(s => s.codec_type === 'video');
        const audioStream = metadata.streams.find(s => s.codec_type === 'audio');

        resolve({
          duration: parseFloat(metadata.format.duration),
          size: metadata.format.size,
          bitrate: metadata.format.bit_rate,
          video: {
            codec: videoStream?.codec_name,
            width: videoStream?.width,
            height: videoStream?.height,
            fps: videoStream?.avg_frame_rate ? eval(videoStream.avg_frame_rate) : config.video.defaultFps,
          },
          audio: audioStream ? {
            codec: audioStream.codec_name,
            sampleRate: audioStream.sample_rate,
            channels: audioStream.channels,
          } : null,
        });
      });
    });
  }

  async splitVideo(videoPath, options = {}) {
    const videoInfo = await this.getVideoInfo(videoPath);
    const segmentDuration = options.segmentDuration || this.options.segmentDuration;
    const outputDir = options.outputDir || this.options.outputDir;
    const baseName = helpers.getFileNameWithoutExtension(videoPath);

    helpers.ensureDirectory(outputDir);

    const segments = helpers.calculateVideoSegments(videoInfo.duration, segmentDuration);
    const outputFiles = [];

    for (const segment of segments) {
      const outputFileName = `${baseName}_segment_${String(segment.index).padStart(3, '0')}.mp4`;
      const outputPath = path.join(outputDir, outputFileName);

      await this._extractSegment(videoPath, outputPath, segment.startTime, segment.duration);

      outputFiles.push({
        ...segment,
        path: outputPath,
        filename: outputFileName,
      });
    }

    return {
      originalVideo: videoPath,
      originalInfo: videoInfo,
      segments: outputFiles,
      segmentCount: outputFiles.length,
    };
  }

  _extractSegment(inputPath, outputPath, startTime, duration) {
    return new Promise((resolve, reject) => {
      const command = ffmpeg(inputPath)
        .setStartTime(startTime)
        .setDuration(duration)
        .output(outputPath)
        .outputOptions([
          '-c:v', 'copy',
          '-c:a', 'copy',
          '-avoid_negative_ts', 'make_zero',
        ]);

      command
        .on('end', () => {
          resolve(outputPath);
        })
        .on('error', (err) => {
          reject(new Error(`视频分割失败: ${err.message}`));
        })
        .run();
    });
  }

  async splitByCustomIntervals(videoPath, intervals, outputDir = null) {
    const videoInfo = await this.getVideoInfo(videoPath);
    const outputDirectory = outputDir || this.options.outputDir;
    const baseName = helpers.getFileNameWithoutExtension(videoPath);

    helpers.ensureDirectory(outputDirectory);

    const outputFiles = [];

    for (let i = 0; i < intervals.length; i++) {
      const interval = intervals[i];
      const startTime = interval.startTime;
      const duration = interval.endTime - interval.startTime;

      const outputFileName = `${baseName}_segment_${String(i).padStart(3, '0')}.mp4`;
      const outputPath = path.join(outputDirectory, outputFileName);

      if (startTime < 0 || startTime + duration > videoInfo.duration) {
        throw new Error(`时间区间超出视频范围: 开始 ${startTime}, 时长 ${duration}, 视频总长 ${videoInfo.duration}`);
      }

      await this._extractSegment(videoPath, outputPath, startTime, duration);

      outputFiles.push({
        index: i,
        startTime,
        endTime: interval.endTime,
        duration,
        path: outputPath,
        filename: outputFileName,
      });
    }

    return {
      originalVideo: videoPath,
      originalInfo: videoInfo,
      segments: outputFiles,
      segmentCount: outputFiles.length,
    };
  }

  async mergeVideos(videoPaths, outputPath, options = {}) {
    return new Promise((resolve, reject) => {
      if (videoPaths.length === 0) {
        reject(new Error('没有提供要合并的视频文件'));
        return;
      }

      const concatFile = this._createConcatFile(videoPaths);

      const command = ffmpeg()
        .input(concatFile)
        .inputOptions(['-f', 'concat', '-safe', '0'])
        .output(outputPath)
        .outputOptions([
          '-c:v', 'copy',
          '-c:a', 'copy',
        ]);

      if (options.videoCodec && options.videoCodec !== 'copy') {
        command.outputOptions('-c:v', options.videoCodec);
      }
      if (options.audioCodec && options.audioCodec !== 'copy') {
        command.outputOptions('-c:a', options.audioCodec);
      }

      command
        .on('end', () => {
          fs.unlinkSync(concatFile);
          resolve(outputPath);
        })
        .on('error', (err) => {
          if (fs.existsSync(concatFile)) {
            fs.unlinkSync(concatFile);
          }
          reject(new Error(`视频合并失败: ${err.message}`));
        })
        .run();
    });
  }

  _createConcatFile(videoPaths) {
    const tempDir = config.directories.temp;
    helpers.ensureDirectory(tempDir);

    const concatFilePath = path.join(tempDir, `concat_${helpers.generateUniqueId()}.txt`);
    const content = videoPaths.map(p => `file '${p.replace(/'/g, "'\\''")}'`).join('\n');

    fs.writeFileSync(concatFilePath, content, 'utf-8');
    return concatFilePath;
  }

  async getAudioVolume(videoPath) {
    return new Promise((resolve, reject) => {
      const command = ffmpeg(videoPath)
        .outputOptions([
          '-af', 'volumedetect',
          '-f', 'null',
          '-vn',
        ])
        .output('-');

      let stderrOutput = '';

      command
        .on('stderr', (stderrLine) => {
          stderrOutput += stderrLine + '\n';
        })
        .on('end', () => {
          const meanVolumeMatch = stderrOutput.match(/mean_volume:\s*(-?[\d.]+)/);
          const maxVolumeMatch = stderrOutput.match(/max_volume:\s*(-?[\d.]+)/);
          
          const meanVolume = meanVolumeMatch ? parseFloat(meanVolumeMatch[1]) : null;
          const maxVolume = maxVolumeMatch ? parseFloat(maxVolumeMatch[1]) : null;

          resolve({
            meanVolume,
            maxVolume,
            hasAudio: meanVolume !== null,
          });
        })
        .on('error', (err) => {
          reject(new Error(`检测音频音量失败: ${err.message}`));
        })
        .run();
    });
  }
}

export default VideoSplitter;